File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public static boolean checkpalindrome(String s){
        int left =0;
        int right=s.length() -1;
        while(left<right){
            char currFirst = s.charAt(left);
        	char currLast = s.charAt(right);
            if(currFirst != currLast){
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    public boolean isPalindrome(String s) {
        StringBuilder sb =new StringBuilder();
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if (Character.isLetter(ch) || Character.isDigit(ch)){
                sb.append(ch);
            }
        }
        String result =sb.toString();
        result = result.toLowerCase();
        if (checkpalindrome(result)){
            return true;
        }
        return false;
    }
}

File: 0125-valid-palindrome/0125-valid-palindrome.cpp
class Solution {
public:
    bool isPal(const string& s, int left, int right) {
        // Move left and right to next alphanumeric character
        while (left < right && !isalnum(s[left])) left++;
        while (left < right && !isalnum(s[right])) right--;

        // Base case
        if (left >= right) return true;

        // Compare characters (case-insensitive)
        if (tolower(s[left]) != tolower(s[right])) return false;

        return isPal(s, left + 1, right - 1); // Recursive step
    }
    bool isPalindrome(string s) {
        return isPal(s, 0, s.length() - 1);   
    }
};